### ISSUE TITLE
Parallel to async waterfall

### ISSUE THREAD

**Issue by LewisJEllis** [caolan/highland#192]
Waterfall is one of the biggest uses of async, and there are some definite parallels to it in Highland:
```javascript
var async = require('async');
var _ = require('highland');

var f1 = function(arg, callback){
  callback(null, arg + 'b');
}

var f2 = function(arg, callback){
  callback(null, arg + 'c');
}

var f3 = function(arg, callback) {
  callback(null, arg + 'd');
}

function chain(s, f) {
  return s.flatMap(_.wrapCallback(f))
}

async.waterfall([async.apply(f1, 'a'), f2, f3], function (err, arg) {
  console.log(arg);
});

_(['a'])
  .flatMap(_.wrapCallback(f1))
  .flatMap(_.wrapCallback(f2))
  .flatMap(_.wrapCallback(f3))
  .apply(_.log);

_([f1, f2, f3])
  .reduce(_(['a']), chain)
  .flatten()
  .apply(_.log);
```
There are probably other ways to do the same thing, too. 

I got to thinking about this from [this blog post](http://blog.vullum.io/javascript-flow-callback-hell-vs-async-vs-highland/) which draws parallels between async and highland, much like `nfcall` does in #171 and like was mentioned in #188.

Thoughts on an explicit `waterfall` parallel in highland? It could implement one of the patterns above, and it might also be able to support passing through multiple arguments. [Async's example](https://github.com/caolan/async#waterfall) does this, but it's a bit of a trick with highland.

**Comment by vqvu** [comment]
I'm not convinced on an explicit `waterfall` method, as "waterfall" (the
word) doesn't really describe a stream transform, imo.

Wouldn't this be better as an entry in a Highland cookbook?

**Comment by jeromew** [comment]
I agree that we should have more cookbooks to show how constructs from other libraries can be ported to highland.

but we should try to keep in the highland core only the core constructs. I think that the next big thing after 3.0 will be to decide something on the extensibility of highland so that an ecosystem could exist. Then there could be an "highland-async" library that has async-like helpers.

**Comment by LewisJEllis** [comment]
The reduce version feels like a stream transform to me, but less obviously so than the `nfcall` counterpart. This is definitely a better fit for a theoretical `highland-async` module (and I think `nfcall` would get broken out into that as well).

In either case, it's definitely cookbook material.

I'll make a label for 'issues to revisit when extensibility gets here'.

**Comment by eiriklv** [comment]
Just to add to the your examples @LewisJEllis - I find myself using composition a lot both for mapping and filtering

Using only sync functions:
```js
_(['a'])
  .map(compose(
    syncf3,
    syncf2,
    syncf1
  ))
  .apply(_.log); // 'abcd'
```

`compose` in this case could be `_.compose` /  `_.flowRight` from Lodash.

Using only async functions:
```js
_(['a'])
  .flatMap(_.wrapCallback(
    async.compose(
      f3,
      f2,
      f1
    )
  ))
  .apply(_.log); // 'abcd'
```

`async.compose` in this case would be from the async library

Or for a mix of sync and async functions:
```js
_(['a'])
  .flatMap(_.wrapCallback(
    async.compose(
      f3,
      asyncify(syncf2),
      f1,
    )
  ))
  .apply(_.log); // 'abcd'
```

where asyncify would be something like:
```js
function asyncify(fn) {
    return function() {
        var error, res;
        var args = Array.prototype.slice.call(arguments);
        var callback = args.pop();

        try {
            res = fn.apply(this, args);
        } catch (e) {
            error = e;
        }

        // use a polyfill here for browser support
        setImmediate(callback.bind(this, error, res)); 
    };
};
```
